lib/Extractor/Solver.cpp

```diff
@@ -76,6 +76,157 @@ class BaseSolver : public Solver {
   BaseSolver(std::unique_ptr<SMTLIBSolver> SMTSolver, unsigned Timeout)
       : SMTSolver(std::move(SMTSolver)), Timeout(Timeout) {}
 
+  bool testZeroMSB(const BlockPCs &BPCs,
+                    const std::vector<InstMapping> &PCs,
+                    Inst *LHS, InstContext &IC) {
+    unsigned W = LHS->Width;
+    Inst *Mask = IC.getConst(APInt::getOneBitSet(W, W-1));
+    InstMapping Mapping(IC.getInst(Inst::And, W, { LHS, Mask }), IC.getConst(APInt::getNullValue(W)));
+    bool IsSat;
+    std::error_code EC = SMTSolver->isSatisfiable(BuildQuery(IC, BPCs, PCs, Mapping, 0, /*Precondition=*/0),
+                                                  IsSat, 0, 0, Timeout);
+    if (EC)
+      llvm::report_fatal_error("stopping due to error");
+    return !IsSat;
+  }
+
+  bool testOneMSB(const BlockPCs &BPCs,
+                   const std::vector<InstMapping> &PCs,
+                   Inst *LHS, InstContext &IC) {
+    unsigned W = LHS->Width;
+    Inst *Mask = IC.getConst(APInt::getOneBitSet(W, W-1));
+    InstMapping Mapping(IC.getInst(Inst::And, W, { LHS, Mask }), Mask);
+    bool IsSat;
+    std::error_code EC = SMTSolver->isSatisfiable(BuildQuery(IC, BPCs, PCs, Mapping, 0, /*Precondition=*/0),
+                                                  IsSat, 0, 0, Timeout);
+    if (EC)
+      llvm::report_fatal_error("stopping due to error");
+    return !IsSat;
+  }
+
+  std::error_code negative(const BlockPCs &BPCs,
+                           const std::vector<InstMapping> &PCs,
+                           Inst *LHS, bool &Negative,
+                           InstContext &IC) override {
+    unsigned W = LHS->Width;
+    Negative = false;
+    if (testZeroMSB(BPCs, PCs, LHS, IC))
```

Comment:
I don't understand this, why are you calling a function and then using its result to set Negative to false, when you just initialized Negative to false one line earlier in the code?